### ISSUE TITLE
Add support for JPEG

### ISSUE THREAD

**Issue by DaddyTheRunner** [lawsie/guizero#33]
How difficult would it be to add support for JPEG and png image formats? Both of these are very common and would open up many more options for kids.

**Comment by lawsie** [comment]
The reason there isn't support for these formats currently is because they would require an additional library installation (PIL/pillow). I wanted to make it as simple as possible for teachers to get guizero installed on their network as often technicians won't install things or can't install things. Not relying on anything that isn't installed by default means that there's an option to install by simply downloading guizero as a zip into a child's home folder - no need even to use pip, and no need to involve technicians.

One option is to support jpeg/png if and only if the required image libraries are present, and if those libraries are not available, to provide a meaningful error message and still support gifs. However I haven't had time to work on doing that.

**Comment by DaddyTheRunner** [comment]
I totally understand your rational for the limited graphics image support.  However, I did a little research and I think the required libraries are pre-installed with the standard Raspbian images.

Is it sufficient to have the `python-pil` and `python3-pil` libraries installed to enable JPEG and png?  If so, according to this [website](https://n8henrie.com/2015/12/default-packages-on-raspbian-jessie-jessie-lite-and-wheezy/) as of 2015, both of these libraries are part of the standard Raspbian image for Wheezy and Jessie, but not Jessie-Lite.  However, I don't think you are targeting Jessie-Light because it also doesn't include `python-tk` which is required for `guizero`.

Therefore, unless I'm missing some other detail in your response, it would appear that the required libraries are already pre-installed with the standard Raspbian images.  What am I possibly missing?

**Comment by lawsie** [comment]
Aha, yes. It's not just aimed at people using it on the Raspberry Pi. Lots of schools won't own Raspberry Pis so will be using it on the computers they already have which are probably PCs, or kids might be using it at home. I've seen several stack overflow questions from people who have been asked to use it for homework, for example! :)

**Comment by DavidWhaleMEF** [comment]
You could write a wrapper class that fails safe (in a degraded but still functional way) if the PIL is not installed. If trying to render a .jpg and PIL is not present, just dump a canvas and draw two red lines from diagonal corners instead.

Just saying, because I might or might not have done something similar recently :)

```try:
    import PIL
except ImportException:
    PIL = None

if PIL is not None:
    PIL.do_jpeg_stuff(tk)
else:
    draw_dummy_canvas(tk)
```

**Comment by martinohanlon** [comment]
Interestingly I just tried to display a PNG file using guizero.Picture and it worked fine.

I will do some testing and update documentation accordingly.

**Comment by martinohanlon** [comment]
Ok, so tkinter (and therefore guizero) seems quite happy opening PNG files, other files types I tried such as JPEG and BMP dont work.

This is probably due to PNG being an evolution of GIF and relatively X compatibility.

I have changed the name of the issue.

**Comment by DavidWhaleMEF** [comment]
I have a recollection this didn't work on Mac when I tried png, so you should probably test on all supported platforms before deciding this is solved, just to make sure. A lot of the other picture types other than GIF tend to be supported by PIL, so if you inadvertently have PIL installed on your system they will work, as the tcl library will find PIL if it exists. I know I don't have PIL installed on my Mac which is probably why it didn't work.

**Comment by martinohanlon** [comment]
Mac testing complete...  MacOS will not load PNG files.

The root cause is the shonky version of Tk for MacOS (default installed and the latest ActiveTCL - other version arent recommended so I didnt try).  I did test with both PIL installed and not installed), same result.

I have tested PNG support for:
 - Windows
 - Ubuntu
 - Nix
 - Raspbian (x86 & Arm)

I will put an OS check in an provide a suitable warning for MacOS users when loading PNGs (probably along the lines of 'this doesnt work due to shonky apple builds')
